dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/DataValueSetImportValidator.java

```diff
@@ -476,12 +467,20 @@ private static void checkDataValuePeriodType( DataValueEntry dataValue, ImportCo
     private static void checkDataValueStrictDataElement( DataValueEntry dataValue, ImportContext context,
         DataSetContext dataSetContext, DataValueContext valueContext )
     {
-        if ( context.isStrictDataElements()
-            && !dataSetContext.getDataSetDataElements().contains( valueContext.getDataElement() ) )
+        if ( !context.isStrictDataElements() )
+        {
+            return;
+        }
+        List<DataSet> targets = dataSetContext.getDataSet() != null
+            ? List.of( dataSetContext.getDataSet() )
+            : context.getValueContextDataSets().get( dataValue.getDataElement(),
```

Comment:
FYI: I go through a cache to make sure this is not loading from DB query more than once per data element. This cache is also used with keys of shape `{de}-{ou}-{aoc}` but using just `{de}` cannot clash and the value type is the same for both.